components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -58,17 +58,77 @@ impl ProofOfAbsence {
 
 /// An enum that represents a proof of existence for a proof list elements.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum ListProof<V> {
+pub enum ProofVariant<V> {
     /// A branch of proof in which both children contain requested elements.
-    Full(Box<ListProof<V>>, Box<ListProof<V>>),
+    Full(Box<ProofVariant<V>>, Box<ProofVariant<V>>),
     /// A branch of proof in which only the left child contains requested elements.
-    Left(Box<ListProof<V>>, Option<Hash>),
+    Left(Box<ProofVariant<V>>, Option<Hash>),
     /// A branch of proof in which only the right child contains requested elements.
-    Right(Hash, Box<ListProof<V>>),
+    Right(Hash, Box<ProofVariant<V>>),
     /// A leaf of the proof with the requested element.
     Leaf(V),
     /// Proof of absence of an element with the specified index.
-    Absent(ProofOfAbsence),
+    Absent(Hash),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ListProof<V> {
+    length: u64,
+    proof: ProofVariant<V>,
+}
+
+impl<V> ListProof<V>
+where
+    V: BinaryValue,
+{
+    pub(crate) fn new(length: u64, proof: ProofVariant<V>) -> Self {
+        Self { length, proof }
+    }
+
+    /// Length of the corresponding ProofList.
+    pub fn length(&self) -> u64 {
+        self.length
+    }
+
+    /// Underlying proof variant.
+    pub fn proof_variant(&self) -> &ProofVariant<V> {
+        &self.proof
+    }
+
+    /// Verifies the correctness of the proof by the trusted list hash and the number of
+    /// elements in the tree.
+    ///
+    /// To validate the proof one needs to provide `expected_list_hash` calculated as follows:
+    /// ```text
+    /// h = sha-256( HashTag::List || len as u64 || merkle_root )
+    /// ```
+    /// and `length` of the `ProofListIndex`.
+    ///
+    /// If the proof is valid, a vector with indices and references to elements is returned.
+    /// Otherwise, `Err` is returned.
+    ///
+    /// If the proof is the proof of absence, then empty vector will be returned.
+    pub fn validate(&self, expected_list_hash: Hash) -> Result<Vec<(u64, &V)>, ListProofError> {
+        let mut vec = Vec::new();
+        let height = self.length.next_power_of_two().trailing_zeros() as u8 + 1;
+        let root_hash =
```

Comment:
is it _root_ hash (= the Merkle root of the tree) or the _list_ hash (the Merkle root combined with the length and the tag)? Below `list_hash` is compared with `root_hash` which does not seem correct.

Reply:
Collect method returns either root_hash and merkle_root. I'll fix that.